platform/external-system-api/src/com/intellij/openapi/externalSystem/model/DataNode.java

```diff
@@ -279,9 +279,13 @@ private void readObject(ObjectInputStream in)
 
   @Override
   public int hashCode() {
-    int result = myChildren.hashCode();
-    result = 31 * result + myKey.hashCode();
+    int result = myKey.hashCode();
     result = 31 * result + getData().hashCode();
+    // We can't use myChildren.hashCode() because it's exponential.
```

Comment:
Yes, the comment is confusing indeed. I think height of a tree is usually 2-3 here, but it may have many nodes.

Why not exclude myChildren from hashcode computation at all? I think nodes with children are project and modules nodes, so it would be enough to compute hashCode of getData(), I doubt that there will be many collisions of ModuleData hashcodes. And this way hashCode for DataNode<ProjectData> won't iterate over all 2000+ module nodes for your project. What do you think?